File: Data Structures & Algorithms/same-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // using recursive DFS
        // if both trees dont exist, they are equal
        if (!p && !q){
            return true;
        }
        // only executes if both trees and subtree comparisons are identical
        if (p && q && p->val == q->val){
            // this compares the subtree of p and q 
            // if the if statement is true, it recurses to check the children and subtrees
            return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
        } else {
            return false;
        }
    }
};

File: Data Structures & Algorithms/invert-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # basic DFS search
        if not root:        # base case
            return None
        
        # swap root's left and right pointers
        root.left, root.right = root.right, root.left

        # recursively call DFS on both left and right child
        self.invertTree(root.left)
        self.invertTree(root.right)

        # return current node (inverted)
        return root